rust/src/mqtt/parser.rs

```diff
@@ -222,99 +222,109 @@ pub fn parse_connect(i: &[u8]) -> IResult<&[u8], MQTTConnectData> {
     ))
 }
 
-pub fn parse_connack(i: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTConnackData> {
-    let (i, topic_name_compression_response) = be_u8(i)?;
-    let (i, return_code) = be_u8(i)?;
-    let (i, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTConnackData {
-            session_present: (topic_name_compression_response & 1) != 0,
-            return_code,
-            properties,
-        },
-    ))
+fn parse_connack(protocol_version: u8) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTConnackData>
+where
+{
+    move |i: &[u8]| {
+        let (i, topic_name_compression_response) = be_u8(i)?;
+        let (i, return_code) = be_u8(i)?;
+        let (i, properties) = parse_properties(i, protocol_version == 5)?;
+        Ok((
+            i,
+            MQTTConnackData {
+                session_present: (topic_name_compression_response & 1) != 0,
+                return_code,
+                properties,
+            },
+        ))
+    }
 }
 
 pub fn parse_publish(
-    i: &[u8], protocol_version: u8, has_id: bool,
-) -> IResult<&[u8], MQTTPublishData> {
-    let (i, topic) = parse_mqtt_string(i)?;
-    let (i, message_id) = cond(has_id, be_u16)(i)?;
-    let (message, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTPublishData {
-            topic,
-            message_id,
-            message: message.to_vec(),
-            properties,
-        },
-    ))
+    protocol_version: u8, has_id: bool,
+) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTPublishData>
+where {
```

Comment:
Is `where` even needed here?  Seems it evaluates to an empty set of constraints.

Reply:
I followed the nom documentation that suggests this syntax to return a closure receiving only a `u8` slice, given a number of parameters. See https://github.com/Geal/nom/blob/main/doc/upgrading_to_nom_5.md#from-macros-to-functions.
I am not familiar with how `where` would otherwise be used but in combination with `move` it does what it needs to do here.